### PR TITLE
Add TS types

### PR DESCRIPTION
**Pull request by orgads**
Copied from DefinitelyTyped.

### CODE REVIEW

**Comment by jankapunkt** [comment]
@orgads I know they to but not everybody uses `express` (which is why there are the adapters). Wouldn't the current import force people to install express?

**Comment by orgads** [comment]
@jankapunkt Done. Hope it's good enough now.

@HappyZombies This is exactly the reason I want to add the definitions file here, so it will be easier to adapt when making changes.

Generally, changes in a minor version should not break existing APIs, so the definitions file is not expected to change frequently.

**Comment by HappyZombies** [comment]
@orgads got it, thanks for explaining, I am not 100% knowledgeable on the whole TypeScript-isms so thanks for being patient with me 😅

**Comment by jankapunkt** [comment]
hey @orgads we are still setting this up, there are many things we have to get running first, before we can merge new features or fixes. Please follow on the following PRs: 

#18 #32 #29 #30 #35

**Comment by orgads** [comment]
ping

**Comment by orgads** [comment]
Changed target to development and rebased.

**Comment by orgads** [comment]
Sorry for bothering, but I'm stuck with a vulnerable lodash dependency because of the original oauth2-server package.

Is anything holding you now from merging this and releasing a new version?

**Comment by jankapunkt** [comment]
@orgads I activate CI for your PR and you need one more review from either @HappyZombies or @jwerre  as I am not the typescript expert

**Comment by orgads** [comment]
Well?

**Comment by Uzlopak** [comment]
If you add typings to your package, then you put them into the typings-attribute of package.json and not into the types-attribute.

**Comment by orgads** [comment]
Not according to [the docs](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html):

> Note that the "typings" field is synonymous with types, and could be used as well.

**Comment by Uzlopak** [comment]
yes you are right. I apologize,

**Comment by jankapunkt** [comment]
@HappyZombies @jwerre I think this is now worth a second look since we released 4.1.0

File: index.d.ts
Comment (Uzlopak) [inline index.d.ts:65]: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9b9cd671114a2a5178809798d8e7f4d8ca6c2671/types/express-serve-static-core/index.d.ts#L375

```suggestion
        constructor(options?: { [key: string]: any } | http.IncomingMessage);
```

**Review by Uzlopak — CHANGES_REQUESTED** [review]
I think http.IncomingMessage is the correct Interface.

File: index.d.ts
Comment (orgads) [reply index.d.ts:65]: Good catch. Thanks!

File: index.d.ts
Comment (Uzlopak) [inline index.d.ts:394]: Does the client not have a secret?

Comment (Uzlopak) [reply index.d.ts:394]: Also should have user_id

Comment (orgads) [reply index.d.ts:394]: I can't find any reference to `client.secret` or `client.client_secret`. `user_id` only appears in spec.rst, but nowhere in the code.

Comment (Uzlopak) [reply index.d.ts:394]: Actually I think the state is fine for now. We should actually merge it and fix stuff in an iterative manner.

Comment (Uzlopak) [reply index.d.ts:394]: RFC 6749 https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1
client_id
client_secret

userId is part of oauth2_server specific implementation

Comment (orgads) [reply index.d.ts:394]: But where is it supported in this library?

Comment (Uzlopak) [reply index.d.ts:394]: Let me think "loud":
You need it if you use confidential clients with ClientCredentialFlow. Then you use the secret in getClient. 

I think you are right: This is up to the implementer.

**Comment by jorenvandeweyer** [comment]
This issue is still keeping me from switching to `@node-oauth/oauth2-server` (and probably many other).

I think this should be merged and released as soon as possible and improved later.

**Comment by orgads** [comment]
I lost my patience and created a forked package with the types: https://www.npmjs.com/package/@audc/oauth2-server

Feel free to use it :)

**Comment by HappyZombies** [comment]
I'll try to get this out tomorrow, there were many issues initially so I was a bit sus, but I'm sure they're ironed out now. Sorry about that y'all.

**Comment by orgads** [comment]
Thank you.

**Review by HappyZombies — APPROVED** [review]

**Comment by HappyZombies** [comment]
@jwerre  mind approving?

**Comment by jorenvandeweyer** [comment]
> Ohhh... I'm not much of a Typescripter. I'll approve but not with any confidence.

I'll verify the types on some existing projects, and update then if necessary.

**Comment by Uzlopak** [comment]
@jwerre 

I am a typescript user and I think that the types look Ok so far. We can always modify them if the are wrong. In worst case a typescript user would do a // @ts-ignore and report this as a finding. 
So nobody should be blocked by the typings and they dont modify active code so no breaking change.

**Comment by HappyZombies** [comment]
100% @Uzlopak , any serious issue I have can be fixed later on. Just wanna unblock people looking forward to this

**Comment by jwerre** [comment]
I wonder it we could set up [DTSLint](https://github.com/microsoft/dtslint) in order to test the types?

**Comment by Uzlopak** [comment]
Wie could use tsd like fastify does.

I added yesterday a type change to fastify/point-of-view and modified the test. 

https://github.com/fastify/point-of-view/pull/272#event-5662740540

**Comment by jorenvandeweyer** [comment]
Thanks for merging! Any idea when this will be released?

**Comment by HappyZombies** [comment]
I'll try to get it out today or tomorrow

**Comment by HappyZombies** [comment]
This has been released on `4.1.1`

**Pull request merged**
